server/metric_reporter.go

```diff
@@ -0,0 +1,130 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/bytebase/bytebase/api"
+	"github.com/bytebase/bytebase/common"
+	enterpriseAPI "github.com/bytebase/bytebase/enterprise/api"
+	"github.com/bytebase/bytebase/plugin/metric/collector"
+	"github.com/bytebase/bytebase/plugin/metric/reporter"
+
+	"go.uber.org/zap"
+)
+
+const (
+	metricSchedulerInterval = time.Duration(1) * time.Hour
+)
+
+// MetricScheduler is the metric scheduler.
+type MetricScheduler struct {
+	l *zap.Logger
+
+	enabled bool
+	// subscription is the pointer to the server.subscription. This can be updated by users so we need to update the identify in each schedule.
+	subscription  *enterpriseAPI.Subscription
+	reporter      reporter.MetricReporter
+	workspaceID   string
+	collectorList []collector.MetricCollector
+}
+
+// NewMetricScheduler creates a new metric scheduler.
+func NewMetricScheduler(logger *zap.Logger, server *Server, workspaceID string) *MetricScheduler {
+	reporter := reporter.NewSegmentReporter(logger, server.profile.SegmentKey, workspaceID)
+	collectorList := []collector.MetricCollector{
+		collector.NewInstanceCollector(logger, server.store),
+		collector.NewIssueCollector(logger, server.store),
+	}
+	enabled := server.profile.Mode == common.ReleaseModeProd && !server.profile.Demo
+
+	return &MetricScheduler{
+		l:             logger,
+		enabled:       enabled,
+		subscription:  &server.subscription,
```

Comment:
We only need to pass the plan here. Also I would rather pass store, segmentkey, plan to NewMetricScheduler instead of passing the whole server. 

Thus the dependency is more clear (e.g. otherwise one would miss server.subscription is required)

Reply:
Oh, I see. Then let's add a comment. It's not that obvious.

Reply:
comment added